Fix #9294: critical bug with template polymorphism (#9918)
coq/coq: doc/changelog/01-kernel/09918-unsound-template-polymorphism.rst

- Declarations of template-polymorphic inductive types ignored the
  provenance of the universes they were abstracting on and did not

>   detect if they should be greater or equal to :math:`Set` in

Reviewer: ```suggestion
  detect if they should be greater or equal to :math:`\Set` in
```

>   detect if they should be greater or equal to Set in general. Previous

Reviewer: Could be:

```suggestion
  detect if they should be greater or equal to :g:`Set` in general. Previous
```

or

```suggestion
  detect if they should be greater or equal to :math:`\Set` in general. Previous
```

  general. Previous universes and universes introduced by the inductive
  definition could have constraints that prevented their instantiation

>   with e.g. :math:`Prop`, resulting in unsound instantiations later. The

Reviewer: ```suggestion
  with e.g. :math:`\Prop`, resulting in unsound instantiations later. The
```

  implemented fix only allows abstraction over universes introduced by
  the inductive declaration, and properly records all their constraints
  by making them by default only `>= Prop`.

  This prevents inductive declarations in section to be universe
  polymorphic over section parameters. For a backward compatible fix,
  simply hoist the inductive definition out of the section.
  An alternative is to declare the inductive as universe-polymorphic and
  cumulative in a universe-polymorphic section: all universes and
  constraints will be properly gathered in this case.
  See :ref:`Template-polymorphism` for a detailed exposition of the
  rules governing template-polymorphic types.

  To help users incrementally fix this issue, a command line option

>   `-no-template-check` and a global option `Unset Template Check` are

Reviewer: ```suggestion
  `-no-template-check` and a global flag :flag:`Template Check` are
```
